lib/base/configobject.cpp

```diff
@@ -413,7 +413,8 @@ void ConfigObject::OnConfigLoaded()
 
 void ConfigObject::OnAllConfigLoaded()
 {
-	static ConfigType *ctype;
+	// Prevent data races without using any mutexes by declaring the variable as a thread_local
+	thread_local static ConfigType *ctype;
```

Comment:
Can anything bad happen at all here?

Reply:
Changing the storage to thread_local won't break anything because it looks like the variable is used for caching purposes only anyways.

I'm not sure why `Zone::GetByName()` isn't just used here? Is this some circular dependency thing? Seems to compile though but haven't tested anything.

Other option would be to just use `static ConfigType *ctype = dynamic_cast<ConfigType *>(Type::GetByName("Zone").get());` (basically the same as the current code but with C++11, the compiler has to ensure it's thread-safe: https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables)

Reply:
> I'm not sure why `Zone::GetByName()` isn't just used here? Is this some circular dependency thing? Seems to compile though but haven't tested anything.

I will also check it out and if you want I can change it!

Reply:
> Is this some circular dependency thing?

That's exactly it! See the following error.

```bash
[ 12%] Building CXX object lib/base/CMakeFiles/base.dir/configtype.cpp.o
[ 12%] Building CXX object lib/base/CMakeFiles/base.dir/configuration.cpp.o
In file included from /Users/yhabteab/Workspace/icinga2/lib/base/configobject.cpp:19:
/Users/yhabteab/Workspace/icinga2/lib/remote/zone.hpp:7:10: fatal error: 'remote/zone-ti.hpp' file not found
#include "remote/zone-ti.hpp"
         ^~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [lib/base/CMakeFiles/base.dir/configobject.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [lib/base/CMakeFiles/base.dir/all] Error 2
make: *** [all] Error 2
```

Reply:
The remote library is depending on the base and not vice versa.

Reply:
Well probably it's something about dependencies between subdirectories of `lib/`. Using it would require including something from `lib/remote/` in `lib/base/` which would only be the second occurrence of such a import and the first one looks like a bug because something was moved from `remote` to `base` in d74b15d6408cb4644e7d9a27ae76d67e7a3dd57f but the include wasn't updated:

```
$ git grep remote/ lib/base/
lib/base/base64.hpp:#include "remote/i2-remote.hpp"
```

But on the other hand, this code has a dependency on `lib/remote/` anyways, if that isn't properly initialized, this function will dereference a null pointer. And then there might be changes needed to the CMake config, so unless you want to go down that rabbit hole, better not use `Zone::GetByName()` here directly :grimacing:

Reply:
> Changing the storage to thread_local won't break anything because it looks like the variable is used for caching purposes only anyways.

No, I mean the original code.

Reply:
> But on the other hand, this code has a dependency on `lib/remote/` anyways

No, it simply includes headers back and forth here, but `lib/base` directory/lib does not depend on the remote lib. That would not work at all!  When the involved libraries are not all static there can' t be any dependency on each other! Only *one* around! However, if there is something that I don't get yet, feel free to explain it to me :)

Reply:
I meant in the sense of a functional dependency: `ConfigObject::OnAllConfigLoaded()` requires `lib/remote/` to be initialized, which registers a type named `Zone` so that `Type::GetByName("Zone")` actually returns something.

Also, is the stuff from `lib/` (apart from `lib/*_shim`) even used as an actual library anywhere? Or is it always all linked together in the big `icinga2` binary anyways? At least my build directory contains no `*.a` or `*.so` files besides `libpgsql_shim.so` and `libmysql_shim.so`.

Reply:
I still see no harm in the original code.

Reply:
> Also, is the stuff from `lib/` (apart from `lib/*_shim`) even used as an actual library anywhere? Or is it always all linked together in the big `icinga2` binary anyways?

They are all linked together into the `icinga2` executable, but there is a dependency order of the subdirectories to control which directory has to be built first, e.g. not to build the remote directory before the base and configuration directories are ready.

Reply:
> They are all linked together into the `icinga2` executable, but there is a dependency order of the subdirectories to control which directory has to be built first, e.g. not to build the remote directory before the `base` and `config` directories are ready.

Yeah, so better stick to that statement:
> And then there might be changes needed to the CMake config, so unless you want to go down that rabbit hole, better not use `Zone::GetByName()` here directly :grimacing:

Reply:
You mean when the pointer 0x123456 gets updated to 0x123456, but only the first $X digits are written, yet? What's the problem then? Oh, wait... you mean if thread $Y has written only $X digits yet, but thread $Z... ?

Reply:
I think it's perfectly fine for a compiler to update a 32 bit pointer from the value `0x00000000` to `0x12345678` by writing two 16 bit values, so that it's first updated to `0x12340000` and then finally to `0x12345678`. If another thread executes the function between the two writes, it might read the invalid pointer `0x12340000` and use it as it's not all zero.

Reply:
![](https://img.welt.de/img/kmpkt/mobile193712903/9312509257-ci102l-w1024/Grumpy-Cat-Makes-Appearance-At-Kitson-Santa-Monica.jpg)